pkg/util/mount/mount_helper_unix.go

```diff
@@ -40,5 +40,5 @@ func IsCorruptedMnt(err error) bool {
 		underlyingError = pe.Err
 	}
 
-	return underlyingError == syscall.ENOTCONN || underlyingError == syscall.ESTALE || underlyingError == syscall.EIO || underlyingError == syscall.EACCES
+	return underlyingError == syscall.ENOTCONN || underlyingError == syscall.ESTALE || underlyingError == syscall.EIO || underlyingError == syscall.EACCES || underlyingError == syscall.EAGAIN
```

Comment:
I would think that `EAGAIN` is not a fatal error, on linux this happens when you are doing non-blocking IO and socket buffer is full(while writing) or there is no data available from socket for read and client must retry.

Reply:
`EAGAIN - APPLICATION_ERROR: "resource temporarily unavailable"`, this happens when SMB server is unavailable, if we don't regard it as corrupted mnt, the `TearDown` func will fail and pod would be in `Terminated` forever.

Reply:
It is possible that samba client is bugged or there is some misconfiguration. Treating EAGAIN here as a fatal error will affect all clients so even if we agree that EAGAIN is a fatal error for samba, all evidence suggests that - it is a transient error and should fix itself.

Reply:
hi @gnufied here the problem is we have only one chance to check this issue in flexvolume TearDown func:
https://github.com/kubernetes/kubernetes/blob/e739b553747940324cf4a91429aea905371f89a1/pkg/volume/flexvolume/unmounter.go#L52-L54
If `EAGAIN` is not a corrupted mnt error, it will goto:
```
return fmt.Errorf("Error checking path: %v", pathErr)
```
And then pod is in `Terminating` state forever

Reply:
Teardown is supposed to retry

Reply:
> Teardown is supposed to retry
While it's always `EAGAIN` error and it's stuck there, I just used the Ubuntu built-in cifs-utils package.

Reply:
I am actually thinking this IsCorruptedMnt check is not necessary. Teardown can just try to execute "umount" and if it fails, we get the errors and also double check whether the mount point still exist or not. Is there anything unsafe to execute "umount" no matter the path is corrupted mount point or not?
@msau42 and I discussed before and we think the teardown function in different plugins should be simplified. Right now, because of this check, we might end up missing some corner cases like this.

Reply:
Among other things, since we have traditionally safe-guarded plugins from unmount call if path does not exist or error is anything other than corrupted mount point,  won't that cause plugins that could not handle such scenarios to fail? cc @chakri-nelluri

Reply:
To be fair, they would have failed anyways.. but like @jingxu97 said, we will have to at minimum check errors and mount point again.

Reply:
Also see this - https://bugzilla.samba.org/show_bug.cgi?id=3857 . EAGAIN should be an internal error not exposed to the user.

Reply:
@gnufied thanks for the check, while I could repro that EAGAIN error easily on my env by changing the SMB server password. Like what @jingxu97 said, we may change the logic of `TeamDown` func, if path error cannot pass `if pathErr != nil && !mount.IsCorruptedMnt(pathErr)`, unmount will never happen which makes the pod in `Terminating` state forever:
https://github.com/kubernetes/kubernetes/blob/e739b553747940324cf4a91429aea905371f89a1/pkg/volume/flexvolume/unmounter.go#L44-L59

Reply:
@gnufied Thinking more we can remove this and have the plugin deal with these errors. Plugins should anyways have to deal with errors and this would one of those errors.

Reply:
ack.I am okay with it too.

Reply:
@chakri-nelluri ok, then let me change the logic of `TearDownAt` in flexvolume to fix this issue.